tests/devices/qubit/test_preprocess.py

```diff
@@ -59,64 +58,116 @@ def has_matrix(self):
+        when only a single layer of expansion is necessary."""
+        op = NoMatOp("a")
+        casted_to_list = list(_operator_decomposition_gen(op))
+        assert len(casted_to_list) == 2
+        assert qml.equal(casted_to_list[0], qml.PauliX("a"))
+        assert qml.equal(casted_to_list[1], qml.PauliY("a"))
+
+    def test_operator_decomposition_gen_decomposed_operator_ragged_nesting(self):
+        """Test that _operator_decomposition_gen handles a decomposition that requires different depths of decomposition."""
+
+        class RaggedDecompositionOp(Operation):
+            num_wires = 1
+
+            def decomposition(self):
+                return [NoMatOp(self.wires), qml.S(self.wires), qml.adjoint(NoMatOp(self.wires))]
+
+        op = RaggedDecompositionOp("a")
+        final_decomp = list(_operator_decomposition_gen(op))
+        assert len(final_decomp) == 5
+        assert qml.equal(final_decomp[0], qml.PauliX("a"))
+        assert qml.equal(final_decomp[1], qml.PauliY("a"))
+        assert qml.equal(final_decomp[2], qml.S("a"))
+        assert qml.equal(final_decomp[3], qml.adjoint(qml.PauliY("a")))
+        assert qml.equal(final_decomp[4], qml.adjoint(qml.PauliX("a")))
+
+    def test_error_from_unsupported_operation(self):
+        """Test that a device error is raised if the operator cant be decomposed and doesn't have a matrix."""
+        op = NoMatNoDecompOp("a")
+        with pytest.raises(DeviceError, match=r"Operator NoMatNoDecompOp"):
+            tuple(_operator_decomposition_gen(op))
+
+
+class TestExpandFnValidation:
+    """Unit tests for helper functions in qml.devices.qubit.preprocess"""
 
-    def test_check_validity_invalid_op(self):
-        """Test that check_validity throws an error when an operation is invalid."""
-        tape = QuantumScript(ops=[qml.Barrier(0)], measurements=[qml.expval(qml.Hadamard(0))])
-        with pytest.raises(DeviceError, match="Gate Barrier not supported on Python Device"):
-            check_validity(tape)
+    def test_error_if_invalid_op(self):
+        """Test that expand_fn throws an error when an operation is does not define a matrix or decomposition."""
+        tape = QuantumScript(ops=[NoMatNoDecompOp(0)], measurements=[qml.expval(qml.Hadamard(0))])
+        with pytest.raises(DeviceError, match="Operator NoMatNoDecompOp"):
+            expand_fn(tape)
 
-    def test_check_validity_invalid_observable(self):
-        """Test that check_validity throws an error when an observable is invalid."""
+    def test_expand_fn_invalid_observable(self):
+        """Test that expand_fn throws an error when an observable is invalid."""
         tape = QuantumScript(
             ops=[qml.PauliX(0)], measurements=[qml.expval(qml.GellMann(wires=0, index=1))]
         )
-        with pytest.raises(DeviceError, match="Observable GellMann not supported on Python Device"):
-            check_validity(tape)
+        with pytest.raises(DeviceError, match=r"Observable GellMann1"):
```

Comment:
```suggestion
        with pytest.raises(DeviceError, match=r"Observable GellMann"):
```

Reply:
```
>>> qml.expval(qml.GellMann(wires=0, index=1))
expval(GellMann1(wires=[0]))
```
Repr for this observable is weird.  I cut it off before the `(` because regex is annoying.

Reply:
That's very odd... I'll add a small PR to fix the repr later